wallet/cardano-trezor-crypto-provider.js

```diff
@@ -0,0 +1,370 @@
+            if (shouldRejectOnError(response.error)) {
+              reject(response.error)
+            }
+          }
+        })
+      })
+    })
+  }
+
+  async function deriveAddress(derivationPath, derivationMode) {
+    const cachedAddress = state.derivedAddresses[JSON.stringify(derivationPath)]
+    if (cachedAddress) {
+      return cachedAddress.address
+    }
+
+    if (derivationMode === 'hardened') {
+      return await _trezorDeriveAddress(derivationPath, true)
+    } else {
+      const xpub = await deriveXpub(derivationPath, derivationMode)
+      const hdPassphrase = new Buffer(await getRootHdPassphrase(), 'hex')
+
+      const address = packAddress(derivationPath, xpub, hdPassphrase)
+      state.derivedAddresses[JSON.stringify(derivationPath)] = {
+        derivationPath,
+        address,
+      }
+
+      return address
+    }
+  }
+
+  async function deriveXpub(derivationPath, derivationMode) {
+    const memoKey = JSON.stringify(derivationPath)
+
+    if (!state.derivedXpubs[memoKey]) {
+      let result
+
+      if (derivationMode === 'hardened') {
+        result = await deriveXpubHardened(derivationPath)
+      } else if (derivationMode === 'nonhardened') {
+        result = await deriveXpubNonHardened(derivationPath)
+      } else {
+        throw Error(`Unknown derivation mode: ${derivationMode}`)
+      }
+
+      state.derivedXpubs[memoKey] = result
+    }
+
+    return state.derivedXpubs[memoKey]
+  }
+
+  async function deriveXpubHardened(derivationPath) {
+    return new Buffer((await deriveTrezorXpub(derivationPath)).xpub, 'hex')
+  }
+
+  async function deriveTrezorXpub(derivationPath) {
+    return await new Promise((resolve, reject) => {
+      // m/44'/1815'/0'/0/childIndex
+      const path = normalizePathsForTrezor([derivationPath])[0]
+
+      callTrezor((shouldRejectOnError) => {
+        TrezorConnect.adaGetXpubKey(path, (response) => {
+          if (response.success) {
+            const xpubData = {
+              xpub: response.xpub_key,
+              root_hd_passphrase: response.root_hd_passphrase,
+            }
+
+            if (!state.rootHdPassphrase) {
+              state.rootHdPassphrase = new Buffer(response.root_hd_passphrase, 'hex')
+            }
+
+            resolve(xpubData)
+          } else {
+            if (shouldRejectOnError(response.error)) {
+              reject(response.error)
+            }
+          }
+        })
+      })
+    })
+  }
+
+  async function deriveXpubNonHardened(derivationPath) {
+    const parentPath = derivationPath.slice(0, derivationPath.length - 1)
+    const childPath = derivationPath.slice(derivationPath.length - 1, derivationPath.length)
+
+    // this reduce ensures that this would work even for empty derivation path
+    return childPath.reduce(derivePublic, await deriveXpub(parentPath, 'hardened'))
+  }
+
+  async function isOwnAddress(address) {
+    const cachedAddress = Object.values(state.derivedAddresses).find((record) => {
```

Comment:
I would simplify this to: `Object.values(state.derivedAddresses).find((record) => record.address === address)`